feat: implement sign arbitrary message

xilosada opened the pull request.

chefsale commented:
cc: @MaximusHaximus @Patrick1904, this is a feature we need for Calimero pretty urgently. I know this week is not the best for you, but appreciate if somebody could take a look to land this asap, as it is blocking us on shipping authentication with Private Shards.

agileurbanite commented:
I agree with @PinkiNice here but let's get some more folks to review.  Adding @MaximusHaximus and @Patrick1904

agileurbanite commented:
Also is there an NEP i can follow to understand what the intent is better?

PinkiNice commented:
@agileurbanite https://github.com/near/near-wallet/issues/2047 this issue describes what we are trying to achieve.

My proposal is to instead of signing provided data, wrap this data into something like
{ 
  message: “Original Message”,
  blockId: “Most recent block hash”,
  accountId: “User account used for signature”,
  publicKey: “Public key used for signature”,
}
And give a signature to this object: this way it cannot be interpreted as a transaction which makes it safe to sign arbitrary data. This approach is cheap to implement and won't require extra transactions for creating FunctionalAK

Patrick1904 commented on packages/frontend/src/routes/SignMessageWrapper.js:
```diff
@@ -0,0 +1,99 @@
+    const dispatch = useDispatch();
+
+    const signCallbackUrl = useSelector(selectAccountUrlCallbackUrl);
+    const meta = useSelector(selectAccountUrlMeta);
```
nit: While not enforced yet, we try to define selectors by simply removing the `select` part. E.g.:

`const accountUrlMeta = useSelector(selectAccountUrlMeta);`

Patrick1904 commented on packages/frontend/src/redux/slices/signMessage/index.js:
```diff
@@ -0,0 +1,81 @@
+    (signMessage) => signMessage.signedRequest || {}
+);
+
+export function addQueryParams(baseUrl, queryParams) {
```
Can we please re-use/re-org this to so that we don't have do duplicate this code from the TX sign slice?

https://github.com/near/near-wallet/blob/master/packages/frontend/src/redux/slices/sign/index.js#L82

Patrick1904 commented on packages/frontend/src/components/sign-message/SignMessage.js:
```diff
@@ -0,0 +1,70 @@
+
+export default ({ sender, message }) => {
+    return (
+        <StyledContainer className="brs-8 bsw-l">
```
nit: can we please use single quotes in jsx

Patrick1904 commented on packages/frontend/src/components/sign-message/SignMessageInvalid.js:
```diff
@@ -0,0 +1,43 @@
+`;
+
+const SignMessageInvalid = () => (
+    <CustomContainer className="small-centered">
```
nit: can we please use single quotes in jsx

xilosada commented:
## Summary

Changes since the last PR:
* Renamed url, components, actions, reducers and slices to `verify-owner`. This was the name was proposed here and is more accurate as this code doesn't sign arbitrary messages.
* Modify signature method to include accountId, blockId, public key and key type to the signature.
* Apply provided design
* Apply feedback
* Wrap `isUrlNotJavascriptProtocol` in a try-catch block. It threw an exception getting the protocol field from an invalid URL.

This PR adds the verify-owner page, which allows sites to request a proof of ownership of the private key of a Near account. For that the site sends a `message` that will be signed along other parameters like. The initial PR allowed to sign arbitrary messages and it could be used by a malicious actor to sign transactions. This new approach uses @agileurbanite suggestions and combines the message with the following fileds:
`
  accountId: user selected near account id,
  message: message generated by the requesting site,
  blockId: hash of the latest block,
  publicKey,
  keyType
`

Page status:
NEEDS_CONFIRMATION(verify-owner-needs-confirmation): Initial status. The user can chose to `confirm` or `reject` the authorization requested. If there is any issue during the signature process a dialog with details of the error is shown and the page will go back to this status.

IN_PROGRESS(verify-owner-in-progress): the verification process is in progress. `Confirm` and `Cancel` buttons are disabled.

COMPLETED(verify-owner-completed): the authentication request has been rejected or completed. The user will be redirected to the `callbackUrl` if it is valid. Otherwise to `/` 


## Test plan
Clone the test page from https://github.com/calimero-is-near/calimero-examples/tree/master/near-sign-message and serve it locally  `python -m SimpleHTTPServer`.

Build and run Near Wallet `yarn && NEAR_WALLET_ENV=testnet yarn start`

Scenarios:
#### Authentication request confirmed:
* Navigate to localhost:8000 and click on the first link.
* Click on `approve`
* Verify that the user is redirected to callback url with the following query arguments:
`
accountId=calitest.testnet
&message=dassdasdasdasadasdas+34+87
&blockId=4TZUNbCGKREqeBAYumtGPBrfyoy36Z98QmSHtYHjuriV
&publicKey=ZHB5f2WZB3Hc08w%2FwfhWsqiSrRL%2BMmQTyx9E%2B9za8r0%3D
&signature=%2FW5drEvUCriBz2XUx6xLkhOmeXazkOckJqArylNdnck6KLMfcotj6fhuoYULx38pJgkGpppPhxCHFLhjvM87Dg%3D%3D
&meta=test
`

https://user-images.githubusercontent.com/2929173/178935822-067f945c-d9dd-4527-b3d0-32ca302e6a0d.mov


#### Authentication request rejected:
* Navigate to localhost:8000 and click on the first link.
* Click on `cancel`
* Verify that the user is redirected to callback url with the following query arguments:
`
meta=test
&error=User+rejected
`

https://user-images.githubusercontent.com/2929173/178936209-35f82c57-7637-4b5b-b0c4-e747508ba583.mov


#### Invalid request:
* Navigate to localhost:8000 and click on the first link.
* Click on approve
* Verify that the invalid request page is shown.

https://user-images.githubusercontent.com/2929173/178936022-ead2db18-35ec-405a-91f8-c97c8ceae57c.mov


#### Auth error:
* Throw an error in `handleAuthorizationRequestConfirmed` packages/frontend/src/redux/slices/verifyOwner/index.js line 40
* Navigate to localhost:8000 and click on the first link.
* Click on approve
* Verify that the error dialog is shown and the wallet returns to the confirmation state.

<img width="1782" alt="Screen Shot 2022-07-14 at 9 06 57 AM" src="https://user-images.githubusercontent.com/2929173/178936289-5d44b4fd-6425-466f-a280-5be09a8cf944.png">

Patrick1904 commented on packages/frontend/src/components/verify-owner/VerifyOwner.js:
```diff
@@ -0,0 +1,109 @@
+        <StyledContainer>
+            <h3><Translate id='verifyOwner.summary.title' /></h3>
+            <div className='logos'>
+                <CalimeroRoundedLogo />
```
I suspect it won't always be Calimero requesting signature here? Might want to opt for something more generic like what's shown on the `/login` confirm screen (`SwapGraphic.js`):

<img width="545" alt="Screen Shot 2022-07-14 at 10 29 01 AM" src="https://user-images.githubusercontent.com/24921205/179006610-d970f4f3-1f48-4b96-9956-3ac7cbdc3d37.png">

chefsale replied:
No, we actually want to be clear it is Calimero in this case.

PinkiNice replied:
@xilosada @chefsale We should stick with a placeholder for now, so using SwapGraphic is the way to go until we come up with a solution to provide customizable icons there.

Patrick1904 commented on packages/frontend/src/components/verify-owner/VerifyOwner.js:
```diff
@@ -0,0 +1,109 @@
+                <NearRoundedLogo/>
+            </div>
+            <h2><Translate id='verifyOwner.summary.connetingWith' data={{accountId}}/></h2>
+            <div className={'account signer'}>
```
nit: can we use regular strings instead of curly's around the strings?
